File: test_comparison/test_bit_order.py
#!/usr/bin/env python3
import imagehash
import numpy as np
from PIL import Image

# Create a simple test pattern
# 8x8 image with predictable pattern
img = Image.new('L', (8, 8), 0)
pixels = np.array(img)

# Set a specific pattern: diagonal line
for i in range(8):
    pixels[i, i] = 255

img = Image.fromarray(pixels)

# Compute hash
ahash = imagehash.average_hash(img)
print(f"Hash: {ahash}")
print(f"Hash hex: {ahash.hash.flatten().tobytes().hex()}")

# Get the bit matrix
bits = ahash.hash.flatten()
print(f"\nBit array (flattened): {bits}")

# Check if it's row-major
print("\nBit matrix (8x8):")
for i in range(8):
    row = ahash.hash[i]
    row_str = ''.join(['1' if x else '0' for x in row])
    print(f"Row {i}: {row_str}")

# Now let's manually compute what we expect
# Average of pixels: (8*255 + 56*0) / 64 = 31.875
# So diagonal pixels > average, others < average
print("\nExpected pattern:")
print("Diagonal should be 1, others 0")
for i in range(8):
    expected = ['1' if j == i else '0' for j in range(8)]
    print(f"Row {i}: {''.join(expected)}")

# Check if Python uses row-major, MSB-first
print("\nChecking bit order...")
# The hash should have bits along the diagonal
hash_hex = str(ahash)
print(f"Hash string: {hash_hex}")

# Convert to binary
hash_int = int(hash_hex, 16)
hash_bin = format(hash_int, '064b')
print(f"Hash binary (64 bits): {hash_bin}")

# Check if diagonal bits are set
print("\nChecking diagonal bits in binary string:")
for i in range(8):
    bit_pos = i * 8 + i  # row-major, MSB first?
    bit_value = hash_bin[bit_pos]
    print(f"  Position {bit_pos} (row {i}, col {i}): {bit_value}")

File: test_comparison/test_bit_order.go
package main

import (
	"fmt"
)

func main() {
	// Test bit ordering
	// Simulate setting bits at positions 63, 62, ..., 0
	var hash uint64 = 0
	
	// Set bit 63 (MSB)
	hash |= 1 << uint(63)
	fmt.Printf("Set bit 63: %064b = 0x%016x\n", hash, hash)
	
	// Set bit 62
	hash |= 1 << uint(62)
	fmt.Printf("Set bit 62: %064b = 0x%016x\n", hash, hash)
	
	// Set bit 1
	hash |= 1 << uint(1)
	fmt.Printf("Set bit 1:  %064b = 0x%016x\n", hash, hash)
	
	// Set bit 0 (LSB)
	hash |= 1 << uint(0)
	fmt.Printf("Set bit 0:  %064b = 0x%016x\n", hash, hash)
	
	// Now test what Python would produce for the same bits
	// Python flattens row-major and converts to hex
	// For a 2x2 array: [[1,0],[0,1]] -> [1,0,0,1] -> binary "1001" -> hex "9"
	// But is "1001" MSB first or LSB first?
	
	fmt.Println("\nPython-style (row-major flattening):")
	// Assume array is [[true, false], [false, true]]
	bits := []bool{true, false, false, true}
	
	// Convert to binary string MSB first
	binaryStr := ""
	for _, b := range bits {
		if b {
			binaryStr += "1"
		} else {
			binaryStr += "0"
		}
	}
	
	fmt.Printf("Binary string (MSB first): %s\n", binaryStr)
	fmt.Printf("As hex: 0x%x\n", uint64(0b1001))
	
	// What if it's LSB first?
	binaryStrLSB := ""
	for i := len(bits)-1; i >= 0; i-- {
		if bits[i] {
			binaryStrLSB += "1"
		} else {
			binaryStrLSB += "0"
		}
	}
	
	fmt.Printf("Binary string (LSB first): %s\n", binaryStrLSB)
	fmt.Printf("As hex: 0x%x\n", uint64(0b1001))
}
